src/streaming/FetchLoader.js

```diff
@@ -0,0 +1,181 @@
+* @param {Object} cfg - dependancies from parent
+*/
+function FetchLoader(cfg) {
+
+    cfg = cfg || {};
+
+    let instance;
+
+    function send(httpRequest) {
+
+        // Variables will be used in the callback functions
+        let firstProgress = true; /*jshint ignore:line*/
+        let needFailureReport = true; /*jshint ignore:line*/
+        let requestStartTime = new Date();
+        let lastTraceTime = requestStartTime; /*jshint ignore:line*/
+        let lastTraceReceivedCount = 0; /*jshint ignore:line*/
+
+        let request = httpRequest.request;
+
+        const headers = new Headers(); /*jshint ignore:line*/
+        if (request.range) {
+            headers.append('Range', 'bytes=' + request.range);
+        }
+
+        if (!request.requestStartDate) {
+            request.requestStartDate = requestStartTime;
+        }
+
+        const reqOptions = {
+            method: httpRequest.method,
+            headers: headers,
+            credentials: httpRequest.withCredentials ? 'include' : undefined
+        };
+
+        fetch(httpRequest.url, reqOptions).then(function (response) {
+            if (!response.ok) {
+                httpRequest.onend();
+            }
+
+            if (!httpRequest.response) {
+                httpRequest.response = {};
+            }
+            httpRequest.response.status = response.status;
+            httpRequest.response.statusText = response.statusText;
+            httpRequest.response.responseURL = response.url;
+            let responseHeaders = '';
+            for (const key of response.headers.keys()) {
+                responseHeaders += key + ': ' + response.headers.get(key) + '\n';
+            }
+            httpRequest.response.responseHeaders = responseHeaders;
+
+            const totalBytes = parseInt(response.headers.get('Content-Length'), 10);
+
+            if (!response.body) {
+                // Fetch returning a ReadableStream response body is not currently supported by all browsers.
+                // Browser compatibility: https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API
+                // If it is not supported, returning the whole segment when it's ready (as xhr)
+                return response.arrayBuffer().then(function (buffer) {
+                    httpRequest.response.response = buffer;
+                    const event = {
+                        loaded: buffer.byteLength,
+                        total: buffer.byteLength
+                    };
+                    httpRequest.progress(event);
+                    httpRequest.onload();
+                    httpRequest.onend();
+                    return;
+                });
+            }
+
+            let bytesReceived = 0;
+            let remaining = new Uint8Array();
+            let dataArray;
+
+            httpRequest.reader = response.body.getReader();
+
+            const processResult = function ({ value, done }) {
```

Comment:
Is all of this strictly necessary? As long as you ensure that bytes are appended to the source buffer in the order in which they are received, MSE is able to do all of this hard work without the need for expensive JS memory manipulation.

Reply:
I am working in some changes related with this part. @bbcrddave, I will let you know once modifications are applied.